### ISSUE TITLE
Commit "Add a Sync call before close when uploading a file to local storage" causes issue with rclone mount

### ISSUE THREAD

**Issue by cvillabrille** [gilbertchen/duplicacy#550]
Hi @gilbertchen

I am testing the new features in duplicacy (as part of this issue found ) and I have realized about something I guess is related to this commit..

I have some remote storages to perform backups set up and available in the machine through rclone mount option (jottacloud, opendrive, hubic) which have stopped working correctly.

Below an example.

this is a test wit the official binary version 2.1.2

rclone mount is as follows:

/usr/bin/rclone mount --allow-other --config /root/.config/rclone/rclone.conf --no-modtime --write-back-cache oDroidOpenDrive:/Test/newapproach /var/duplicacy --ignorchecksum --write-back-cache -v

rclone version is: v1.46.0-028-g972e27a8-beta

now the example

\# duplicacy_linux_arm64_2.1.2  -v  backup -stats  -threads 30  -storage openDrive_rclone_dummy
Storage set to /var/duplicacy
Enter storage password:*********************
Using 16384 iterations for key derivation
Downloading latest revision for snapshot dataDroid
Listing revisions for snapshot dataDroid
No previous backup found
Indexing /data/Downloads/duplicacy_test
Packing .jottaignore
Use 30 uploading threads
Packed .jottaignore (0)
Packing 11-Everything_Will_Be_Alright.mp3
Packed 11-Everything_Will_Be_Alright.mp3 (8273920)
Packing 128mb_partition_5tb.img
Packed 128mb_partition_5tb.img (10715136)
Packing GOPR0026-3.MP4
Uploaded chunk 3 size 9951198, 1.90MB/s 00:47:09 0.1%
Uploaded chunk 1 size 12468224, 4.28MB/s 00:20:53 0.3%
Uploaded chunk 2 size 4194304, 3.63MB/s 00:24:37 0.4%
Uploaded chunk 5 size 11805384, 4.58MB/s 00:19:27 0.6%
Uploaded chunk 4 size 16777216, 6.58MB/s 00:13:30 0.9%
Uploaded chunk 7 size 6655744, 6.55MB/s 00:13:32 1.0%
Uploaded chunk 8 size 5183438, 7.10MB/s 00:12:29 1.1%
Uploaded chunk 9 size 5396789, 6.91MB/s 00:12:49 1.2%
Uploaded chunk 6 size 14622081, 7.55MB/s 00:11:42 1.5%
Uploaded chunk 11 size 6824305, 7.46MB/s 00:11:49 1.6%
Uploaded chunk 10 size 8638472, 8.15MB/s 00:10:49 1.8%
Uploaded chunk 13 size 6217764, 8.64MB/s 00:10:11 1.9%
Uploaded chunk 12 size 11235989, 8.80MB/s 00:09:59 2.1%
Uploaded chunk 15 size 6718828, 8.63MB/s 00:10:10 2.2%
Uploaded chunk 14 size 11752378, 9.43MB/s 00:09:17 2.4%
Uploaded chunk 17 size 10530192, 8.88MB/s 00:09:50 2.6%
Uploaded chunk 19 size 4398598, 9.14MB/s 00:09:33 2.7%
Uploaded chunk 18 size 5183317, 9.45MB/s 00:09:14 2.8%
Uploaded chunk 16 size 11305248, 10.12MB/s 00:08:36 3.0%
Uploaded chunk 21 size 7393608, 8.90MB/s 00:09:46 3.1%
Uploaded chunk 20 size 16777216, 9.25MB/s 00:09:22 3.4%
Uploaded chunk 22 size 16777216, 9.14MB/s 00:09:27 3.7%
Uploaded chunk 24 size 8503649, 9.51MB/s 00:09:04 3.8%
Uploaded chunk 23 size 15075543, 9.72MB/s 00:08:51 4.1%
^CIncomplete snapshot saved to /data/Downloads/duplicacy_test/.duplicacy/incomplete
then I repeat the test but now with the new compiled binary after cleaning up the folder on the remote (openDrive in this case)

\# duplicacy_linux_arm64_2.1.2-webdav  -v -debug   backup -stats  -threads 30  -storage openDrive_rclone_dummy
Storage set to /var/duplicacy
Reading the environment variable DUPLICACY_OPENDRIVE_RCLONE_DUMMY_PASSWORD
Failed to get the value from the keyring: Unable to open dbus session &{%!s(*dbus.Conn=&{0x4000218a60 0x400017b650 true 73d7dbe9bb64ae492c90730c5c77dfd4 [:1.40] {{0 0} 0 0 0 0} {0 0} 3 map[0:true] map[] {{0 0} 0 0 0 0} map[] {{0 0} 0 0 0 0} 0x40000bec00 false {{0 0} 0 0 0 0} [] {0 0} <nil> {0 0}}) org.freedesktop.secrets /org/freedesktop/secrets}: The name org.freedesktop.secrets was not provided by any .service files

Enter storage password:*********************
Using 16384 iterations for key derivation
Chunk read levels: [1], write level: 1
Compression level: 100
Average chunk size: 8388608
Maximum chunk size: 16777216
Minimum chunk size: 4194304
Chunk seed: 7f9db2e8ed1b5e9e592896b10d34dce92e731eb9716678e0f17d726ad5c497da
Reading the environment variable DUPLICACY_OPENDRIVE_RCLONE_DUMMY_PASSWORD
Failed to store the value to the keyring: Unable to open dbus session &{%!s(*dbus.Conn=&{0x4000218a60 0x400017b650 true 73d7dbe9bb64ae492c90730c5c77dfd4 [:1.40] {{0 0} 0 0 0 0} {0 0} 3 map[0:true] map[] {{0 0} 0 0 0 0} map[] {{0 0} 0 0 0 0} 0x40000bec00 false {{0 0} 0 0 0 0} [] {0 0} <nil> {0 0}}) org.freedesktop.secrets /org/freedesktop/secrets}: The name org.freedesktop.secrets was not provided by any .service files

top: /data/Downloads/duplicacy_test, quick: true, tag:
Downloading latest revision for snapshot dataDroid
Listing revisions for snapshot dataDroid
No previous backup found
Indexing /data/Downloads/duplicacy_test
Listing
Incomplete snapshot loaded from /data/Downloads/duplicacy_test/.duplicacy/incomplete
Listing all chunks
Listing chunks/
The incomplete snapshot contains 3 files and 4 chunks
Last chunk in the incomplete snapshot that exist in the storage: -1
Skipped 0 files from previous incomplete backup
Packing 11-Everything_Will_Be_Alright.mp3
Use 30 uploading threads
Packed 11-Everything_Will_Be_Alright.mp3 (8273920)
Packing 128mb_partition_5tb.img
Packed 128mb_partition_5tb.img (10715136)
Packing GOPR0026-3.MP4
Failed to upload the chunk ef1ebab0d04472c3bee7580cddfecbc6f4a5793da6eb0d1d232df152affc3de7: sync /var/duplicacy/chunks/ef/1ebab0d04472c3bee7580cddfecbc6f4a5793da6eb0d1d232df152affc3de7.rgazjjqv.tmp: input/output error
Failed to upload the chunk 3ac6e484b3f234941786f4c8a517deaa9a48c19b9a307e5a3c9324b4fa8877fa: sync /var/duplicacy/chunks/3a/c6e484b3f234941786f4c8a517deaa9a48c19b9a307e5a3c9324b4fa8877fa.iotwdgqq.tmp: input/output error
Failed to upload the chunk 37c75616bfe19eb071c84c126b6ff9316f097ebfb69a1648644c722f8c8f93d3: sync /var/duplicacy/chunks/37/c75616bfe19eb071c84c126b6ff9316f097ebfb69a1648644c722f8c8f93d3.jwiypugc.tmp: input/output error
Failed to upload the chunk 310887604ef86f0d6e7684b284da4d237bc50a1486ff082b2875aadce7e4486c: sync /var/duplicacy/chunks/31/0887604ef86f0d6e7684b284da4d237bc50a1486ff082b2875aadce7e4486c.azhdqafz.tmp: input/output error
Failed to upload the chunk 73a8a4dec3534221af66ce2a27bece3aa0a063b2d2df33bffc30e6f88432da09: sync /var/duplicacy/chunks/73/a8a4dec3534221af66ce2a27bece3aa0a063b2d2df33bffc30e6f88432da09.lrecqdfs.tmp: input/output error
Failed to upload the chunk 9bd3364dd12cb465cafb439958bc80486c7083620761a4e5d83d17ca52185cac: sync /var/duplicacy/chunks/9b/d3364dd12cb465cafb439958bc80486c7083620761a4e5d83d17ca52185cac.bwnqikqw.tmp: input/output error
Chunk 8501ebe79f3162f5d4e890dc23de457f0519cc2c2057d0f0dbf41eadaa3d60bd has been uploaded
Uploaded chunk 2 size 4194304, 683KB/s 02:14:22 0.0%
Failed to upload the chunk ab79da8a53fad6e625885c825144cc425d14a6060b5f080a8e2dfc940712d3ab: sync /var/duplicacy/chunks/ab/79da8a53fad6e625885c825144cc425d14a6060b5f080a8e2dfc940712d3ab.ycyuodav.tmp: input/output error
Failed to upload the chunk 57439e439941e3ef474883b90b7e99e320817d219aa24b583787e7758a91a79a: sync /var/duplicacy/chunks/57/439e439941e3ef474883b90b7e99e320817d219aa24b583787e7758a91a79a.alectqnr.tmp: input/output error
Failed to upload the chunk 91dc9e8f3cba3f2f95cedfe7401309a9f8ba63de33a8c64b108e88e52ac1e3fe: sync /var/duplicacy/chunks/91/dc9e8f3cba3f2f95cedfe7401309a9f8ba63de33a8c64b108e88e52ac1e3fe.amryeqbu.tmp: input/output error
Failed to upload the chunk 9274beb5e424bc9aaf7842c8243cb0b6093c5f5ce3030bd5af3a6042ddd97b28: sync /var/duplicacy/chunks/92/74beb5e424bc9aaf7842c8243cb0b6093c5f5ce3030bd5af3a6042ddd97b28.fwbchebf.tmp: input/output error
Incomplete snapshot saved to /data/Downloads/duplicacy_test/.duplicacy/incomplete
After the first file uploaded, it fails. I already tried many times with same result.

File takes some time to be uploaded from rclone side, but is finally uploaded...

2019/02/28 15:40:49 INFO : chunks/ef/1ebab0d04472c3bee7580cddfecbc6f4a5793da6eb0d1d232df152affc3de7.rgazjjqv.tmp: Copied (new) 2019/02/28 15:40:49 INFO : chunks/3a/c6e484b3f234941786f4c8a517deaa9a48c19b9a307e5a3c9324b4fa8877fa.iotwdgqq.tmp: Copied (new) 2019/02/28 15:40:50 INFO : chunks/31/0887604ef86f0d6e7684b284da4d237bc50a1486ff082b2875aadce7e4486c.azhdqafz.tmp: Copied (new) 2019/02/28 15:40:50 INFO : chunks/ab/79da8a53fad6e625885c825144cc425d14a6060b5f080a8e2dfc940712d3ab.ycyuodav.tmp: Copied (new) 2019/02/28 15:40:51 INFO : chunks/91/dc9e8f3cba3f2f95cedfe7401309a9f8ba63de33a8c64b108e88e52ac1e3fe.amryeqbu.tmp: Copied (new) 2019/02/28 15:40:51 INFO : chunks/73/a8a4dec3534221af66ce2a27bece3aa0a063b2d2df33bffc30e6f88432da09.lrecqdfs.tmp: Copied (new) 2019/02/28 15:40:51 INFO : chunks/57/439e439941e3ef474883b90b7e99e320817d219aa24b583787e7758a91a79a.alectqnr.tmp: Copied (new) 2019/02/28 15:40:51 INFO : chunks/9b/d3364dd12cb465cafb439958bc80486c7083620761a4e5d83d17ca52185cac.bwnqikqw.tmp: Copied (new) 2019/02/28 15:40:51 INFO : chunks/92/74beb5e424bc9aaf7842c8243cb0b6093c5f5ce3030bd5af3a6042ddd97b28.fwbchebf.tmp: Copied (new)

Looks like not that fast as duplicacy expects.

I have tried enabling cache in rclone side and all the options, but from the new commit since 2.1.2, I can only guess this is related to the issue.

**Comment by cvillabrille** [comment]
I have updated the code... within the file duplicacy/src/duplicacy_filestorage.go

new binary file is

-rwxr-xr-x 1 root root 27666616 feb 28 16:13 /mnt/cache/duplicacy_arm/src/github.com/gilbertchen/duplicacy/duplicacy_main

I have commented out below code removing the new Sync included...

       reader := CreateRateLimitedReader(content, storage.UploadRateLimit/storage.numberOfThreads)
        _, err = io.Copy(file, reader)
        if err != nil {
                file.Close()
                return err
        }

        //err = file.Sync()
        //if err != nil {
        //      file.Close()
              return err
        //}

Most probably I have commented out not only the sync...

but I have returned to upload normally.

# /mnt/cache/duplicacy_arm/src/github.com/gilbertchen/duplicacy/duplicacy_main -v backup -stats -threads 30 -storage "openDrive_rclone_dummy" Storage set to /var/duplicacy Enter storage password:********************* Using 16384 iterations for key derivation [...]

I cancelled it when it was in 30% when before I could not even reach 2%...

Uploaded chunk 165 size 4950283, 8.20MB/s 00:07:37 30.3%


I have had it running for 4 days and with the core removed, I have to say I've ever had such good performance with rclone mount. Normally it was failing after some time... now I can upload normally with almost 0% of failures.

**Comment by cvillabrille** [comment]
Hi @gilbertchen 
In addition to remove the sync, could we consider the following improvement?

I guess one possible solution for this issue could be avoid renaming files when we are working in local folders?

I mount remote folder from storage providers through rclone as local (rclone mount) and this temporary renaming causes a lot of issues (cancel any upload due to errors renaming files when it is running for long time the backup).

we can find the temporary file naming here...

temporaryFile := fullPath + "." + string(suffix) + ".tmp"


and used here

// MoveFile renames the file.
func (storage *FileStorage) MoveFile(threadIndex int, from string, to string) (err error) {
        return os.Rename(path.Join(storage.storageDir, from), path.Join(storage.storageDir, to))
}

and here

err = os.Rename(temporaryFile, fullPath)


Can we avoid renaming files in local when working with local storage? That would avoid the issues found when mounting locally remote folders through rclone.

**Comment by cvillabrille** [comment]
Well, another update which promises a better approach.

State of the art: 
* duplicacy makes backup to 5 remote storage servers through rclone mount.
* duplicacy upload the filenames with .tmp extension and then rename it when sync.
* duplicacy fails from time to time due to it tries to check a file name yet uploading with .tmp extension but it looks up for it without .tmp extension (it does not exist yet). I will neither blame duplicacy nor rclone, does not matter, the question are the continue errors)
* every time you have to re-run the backup it takes hours to check on all already uploaded chunks and then continue from last valid chunk.
* At the end, every backup takes one week or more due to the continues errors/retries. 

Solution found tested and confirmed (for me, use it at your own, remember always the programmer is right... ;-)

After realize backups were uploading files with the temporary name, but not renaming due to I have removed the file.Close() call, I have now a better approach that has been working since the last 36hours continuously with no cancellations or error at all (this is the longest time running a backup without any error).

* I have tested all the changes with 1Gb data uploaded (backup) and downloaded (restore) within duplicacy with rclone serving all remote storage. 

* I can confirm it works for me using rclone mount.
```
oDroid-Jottaold:Duplicacy/dataDroid         1,0P    11T  1,0P   1% /mnt/JottaCloud/Duplicacy/dataDroid
oDroid-oneDriveB:Duplicacy/dataDroid        5,0T   1,1T  4,0T  21% /mnt/OneDrive/Duplicacy/dataDroid
oDroidOpenDriveNative:Duplicacy/dataDroid   1,0P      0  1,0P   0% /mnt/OpenDrive/Duplicacy/dataDroid
oDroidB2:oDroidB2-data                      1,0P      0  1,0P   0% /mnt/BackBlaze/Duplicacy/dataDroid
oDroidHub:default                           1,0P   6,9T  1,0P   1% /mnt/Hubic/Duplicacy/dataDroid
```
 

* I have tested this change in 6 different storage locations as local disk, jottaCloud, openDrive, oneDrive business, backBlaze, hubic (I hate hubic!!!) having successful results (which does not mean will work for everyone).

Let me explain the changes.

in your GOPATH I have update the following (kept commented the original code for reference):

src/github.com/gilbertchen/duplicacy/src/duplicacy_filestorage.go
 
(keep always a backup of this file!!!!)

below procedure creates a temporary file name when uploading the file locally based on adding 8 random chars + .tmp extension, to later on rename the file and remove the chars and .tmp extension. So I keep the temporaryFile as fullPath (original name)

```
    temporaryFile := fullPath
        //temporaryFile := fullPath + "." + string(suffix) + ".tmp"
```


Returned back the change introduced after[ this commit (add a sync call before close)](https://github.com/gilbertchen/duplicacy/commit/bb652d0a8c7e89e9f0d4cb8acef7fdc453e21a6b) and let only the original call to file.Close:

```
//      err = file.Sync()
//      if err != nil {
//              file.Close()
//            return err
//      }
//
//      err = file.Close()
         file.Close()
//      if err != nil {
//              return err
//      }
```


below, there is another routine that renames the file from filename.8randchar.tmp to filename, I do not need it as it was causing errors on rclone due to the renaming action when the file was not yet uploaded.

Comment below lines:

```
//      err = os.Rename(temporaryFile, fullPath)
//      if err != nil {
//
//              if _, e := os.Stat(fullPath); e == nil {
//                      os.Remove(temporaryFile)
//                      return nil
//              } else {
//                      return err
//              }
//      }
        return nil
}
```
(remember to keep "return nil" at the end)

compile again :

` go build duplicacy/duplicacy_main.go`

Test it with:
 `$GOPATH/src/github.com/gilbertchen/duplicacy/duplicacy_main backup .....`

you will have, as each time you compile the new binary...

```
VERSION:
   2.1.2 (unofficial)
```



Summarizing changes.... after line 180 in duplicacy_filestore.go file, file must look like:

```
        temporaryFile := fullPath

        file, err := os.OpenFile(temporaryFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
        if err != nil {
                return err
        }

        reader := CreateRateLimitedReader(content, storage.UploadRateLimit/storage.numberOfThreads)
        _, err = io.Copy(file, reader)
        if err != nil {
                file.Close()
                return err
        }

         file.Close()

        return nil
}
```


I really appreciate duplicacy work ot make it as much reliable as possible however, and based in the kind of backups I do in remote based on rclone mount, I do require to perform those changes to ensure there is no issues on uploading. 

If something goes wrong, each time duplicacy has to restart, it has to check all the chunks and with 2Tb of data to be backed up it takes hours only on checks, and it was repeating this check every some time when it was repeatedly failing due to renaming actions not performed.

**Comment by cvillabrille** [comment]
Final comment out the above chance.

Output of backup script after almost 40 hours running with 0 errors... before it was failing randomly every some hours after an attempt to rename a file after some hours...

Rclone mount options:

>  /usr/bin/rclone mount --allow-other --config /root/.config/rclone/rclone.conf --checksum --no-modtime oDroid-oneDriveB:Duplicacy/dataDroid /mnt/OneDrive/Duplicacy/dataDroid --log-file /mnt/cache/log/mount.oneDrive -v


duplicacy backup command (within a script):

>  duplicacy -v -comment rclone_oneDrive backup -stats  -threads 30   -storage oneDrive_rclone_dummy 


Duplicacy check command:

> duplicacy check -a -tabular -stats -storage oneDrive_rclone_dummy 


Final result with 0 errors from scratch until the end (almost 40 hours uploading 1Tb of data)
```
Listing snapshots/
Listing snapshots/dataDroid/
Listing chunks/
Backup for /data at revision 1 completed
Removed incomplete snapshot /data/.duplicacy/incomplete
Files: 74806 total, 1091G bytes; 74806 new, 1091G bytes
File chunks: 117551 total, 1091G bytes; 96484 new, 998,208M bytes, 710,593M bytes uploaded
Metadata chunks: 5 total, 39,549K bytes; 5 new, 39,549K bytes, 13,528K bytes uploaded
All chunks: 117556 total, 1091G bytes; 96489 new, 998,247M bytes, 710,606M bytes uploaded
Total running time: 1 day 16:16:51
@@@@@@@@@@@@@@@@@@@@@@@@@@@@ Backup succesful at Wed Mar 6 19:24:18 CET 2019 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@
waiting 10 mins to start the check and allow any remaining upload from rclone
 ############## Starting backup check at  Wed March 6 19:34:18 CET 2019 #############
Storage set to /mnt/OneDrive/Duplicacy/dataDroid
Listing all chunks
1 snapshots and 1 revisions
Total chunk size is 710,606M in 96489 chunks
All chunks referenced by snapshot dataDroid at revision 1 exist

      snap | rev |                          | files | bytes | chunks |    bytes |  uniq |    bytes |   new |    bytes |
 dataDroid |   1 | @ 2019-03-05 03:04 -hash | 74806 | 1091G |  96489 | 710,606M | 96489 | 710,606M | 96489 | 710,606M |
 dataDroid | all |                          |       |       |  96489 | 710,606M | 96489 | 710,606M |       |          |

```

Tested restoring data to ensure the backup was correct. All ok.

Can you please take into consideration a review on the temporary file names renaming or introduce a check on whether the folder is remote rclone mountpoint ( I cannot imagine how, however I am sure there are options for doing so. 

FYI: mount point uses type fuse.rclone (below mount output):

`oDroid-oneDriveB:Duplicacy/dataDroid on /mnt/OneDrive/Duplicacy/dataDroid type fuse.rclone (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other)`


Thanks in advance for such wonderful tool. You know you have all my support.

You rock!!

**Comment by cvillabrille** [comment]
I am sorry for the bit delay on my reply.

I did something even better :)

I have captured two situations for you.

1) Mounting a remote openDrive share on local and try to backup (commands and output are in the zip file). Final result is cancel, and upload speed from duplicacy is 1.8 mb... , real network statistics shows more or less the same.
If you try to backup with duplicacy in local, it creates a file with 8 random chars + .tmp as name and then renames this file when it is written. that causes when using rclone mount a lot of failures and retries, [as it is explained here](https://github.com/gilbertchen/duplicacy/issues/550#issuecomment-470347330)... I have managed to change the duplicacy code (OMG so sorry about that!) and I have removed this renaming procedure which has made a backup of 24 hours to work continuously. For this test I have used the compiled binary without this removal. Maybe you can investigate on that to improve the backups mounted through rclone in local.
[log_rclone_mount.zip](https://github.com/gilbertchen/duplicacy/files/2982837/log_rclone_mount.zip)


2) Sharing through rclone serve webdav and performing the same backup test.
[log_rclone_webdav.zip](https://github.com/gilbertchen/duplicacy/files/2982919/log_rclone_webdav.zip)
The size of logs is very different, bigger on the "local" backup.
Backup lasted for 12 mins with webdav, when it is suppose to take 1 or 2 mins to upload 1Gb at full speed (200Mb upload speed bandwidth)


Test was made with 1Gb of data, cache folder is in a SSD disk, data for backup is in spinning disk.

I have noticed also openDrive speed is very low compared with others, does not matter if its through rclone native or webdav option (then you have the limitation of connections which must be less than 30 concurrent connections, otherwise you receive webdav errors and no upload is performed).

**Comment by cvillabrille** [comment]
Hi @gilbertchen, I am updating this topic and I am going to proceed to close it. 

I have to surrender to your expertise since, after the latest version you deployed (2.2.0), I have no more any issue with both webdav (rclone serve webdav) or rclone mount ( I have made a lot of tests), hence I proceed to remove all instructions to tweak the code as it is not really needed. 

I am very proud of you and your team, and this is just another reason to continue supporting you. 

I do not really need it (I use only the CLI version) but I am going to buy another license extension for the GUI to continue supporting you, in addition to the 5 years license I already acquired.

Congratulations, you have made a very good work. Continue doing such a good work. You Rock!!
